core/src/main/scala/tech/beshu/ror/api/ConfigApi.scala

```diff
@@ -133,27 +109,18 @@ class ConfigApi(rorInstance: RorInstance,
 
   private def rorConfigFrom(payload: String) = {
     for {
-      json <- EitherT.fromEither[Task](io.circe.parser.parse(payload).left.map(_ => Failure("JSON body malformed")))
-      rorConfig <- settingsValue(json)
+      request <- EitherT.fromEither[Task](decodeUpdateRequest(payload))
+      rorConfig <- testConfig(request.configString)
     } yield rorConfig
   }
 
-  private def settingsValue(json: io.circe.Json) = {
-    def liftFailure(failureMessage: String) = EitherT.leftT[Task, RawRorConfig](Failure(failureMessage))
-
-    json \\ "settings" match {
-      case Nil =>
-        liftFailure("Malformed settings payload - no settings key")
-      case configJsonValue :: Nil =>
-        configJsonValue.asString match {
-          case Some(configString) =>
-            EitherT(RawRorConfig.fromString(configString).map(_.left.map(error => Failure(error.show))))
-          case None =>
-            liftFailure("Malformed settings payload - settings key value is not string")
-        }
-      case _ =>
-        liftFailure("Malformed settings payload - only one settings value allowed")
-    }
+  private def decodeUpdateRequest(payload: String): Either[Failure, UpdateConfigRequest] = {
+    io.circe.parser.decode[UpdateConfigRequest](payload)
+      .left.map(error => Failure(s"JSON body malformed: [${error.getMessage}]"))
```

Comment:
I'm not sure if plain circe error is enough for this message. We can make pretty message with failed cursor (path).